scripts/k3s-versions.go

```diff
@@ -0,0 +1,189 @@
+			continue
+		case semver.IsValid(channel.Latest) && semver.Compare(channel.Latest, minimumVersion) >= 0:
+			break
+		default:
+			continue
+		}
+		// Turn "v1.31.3+k3s1" into "1.31.3"
+		latest := strings.TrimPrefix(channel.Latest, "v")
+		latest = strings.SplitN(latest, "+", 2)[0]
+		k3sChannels[channel.Name] = latest
+	}
+
+	return k3sChannels, nil
+}
+
+type GithubRelease struct {
+	TagName    string `json:"tag_name"`
+	Draft      bool   `json:"draft"`
+	Prerelease bool   `json:"prerelease"`
+}
+
+// getGithubReleasesPage fetches a single page of GitHub releases and returns a list
+// of all non-draft, non-prerelease releases higher than the minimumVersion.
+func getGithubReleasesPage(page int) ([]GithubRelease, error) {
+	url := fmt.Sprintf("https://api.github.com/repos/k3s-io/k3s/releases?page=%d", page)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	token := os.Getenv("GH_TOKEN")
+	if token == "" {
+		token = os.Getenv("GITHUB_TOKEN")
+	}
+	if token != "" {
+		req.Header.Set("Authorization", "token "+token)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		//nolint:revive // error-strings
+		return nil, fmt.Errorf("GitHub API request failed with status: %s", resp.Status)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var releases []GithubRelease
+	if err := json.Unmarshal(body, &releases); err != nil {
+		return nil, err
+	}
+
+	// Filter desired releases here, so caller will stop requesting additional pages if there are
+	// no more matches (heuristics, but releases are returned in reverse chronological order).
+	releases = slices.DeleteFunc(releases, func(release GithubRelease) bool {
+		return release.Draft || release.Prerelease || semver.Compare(release.TagName, minimumVersion) < 0
+	})
+	return releases, nil
+}
+
+// getGithubReleases returns a sorted list of all matching GitHub releases.
+func getGithubReleases() ([]string, error) {
+	releaseMap := make(map[string]string)
+	for page := 1; ; page++ {
+		releases, err := getGithubReleasesPage(page)
+		if err != nil {
+			return nil, err
+		}
+		if len(releases) == 0 {
+			break
+		}
+		for _, release := range releases {
+			version := semver.Canonical(release.TagName)
+			// for each version we only keep the latest k3s patch, i.e. +k3s2 instead of +k3s1
+			if oldTag, ok := releaseMap[version]; ok {
+				oldPatch, _ := strconv.Atoi(strings.TrimPrefix(semver.Build(oldTag), "+k3s"))
+				patch, _ := strconv.Atoi(strings.TrimPrefix(semver.Build(release.TagName), "+k3s"))
+				if oldPatch > patch {
+					continue
+				}
+			}
+			releaseMap[version] = release.TagName
+		}
+	}
+
+	return slices.SortedFunc(maps.Values(releaseMap), semver.Compare), nil
+}
+
+func getK3sVersions() (string, error) {
+	if len(os.Args) > 1 {
+		minimumVersion = os.Args[1]
+	}
```

Comment:
Do we still need this? Without it `minimumVersion` can be a `const`, I think. And we no longer need to check its validity.

Reply:
We don't really need it (and it was originally a `const`). I only added it because while testing I ran it like this:

```console
❯ go run scripts/k3s-versions.go v1.30
{
  "cacheVersion": 2,
  "channels": {
    "latest": "1.31.4",
    "stable": "1.31.4",
    "v1.30": "1.30.8",
    "v1.31": "1.31.4"
  },
  "versions": [
    "v1.30.0+k3s1",
    "v1.30.1+k3s1",
    "v1.30.2+k3s2",
    "v1.30.3+k3s1",
    "v1.30.4+k3s1",
    "v1.30.5+k3s1",
    "v1.30.6+k3s1",
    "v1.30.7+k3s1",
    "v1.30.8+k3s1",
    "v1.31.0+k3s1",
    "v1.31.1+k3s1",
    "v1.31.2+k3s1",
    "v1.31.3+k3s1",
    "v1.31.4+k3s1"
  ]
}
```

But I guess this is no longer a valid use case, so I will remove it.